### PR TITLE
MONAI integration, part 1 of 4: handler lifecycle, step and Input Wait

### PR DESCRIPTION
**Pull request by Pendu**
Part 1 of 4 for MONAI support (#464), planned in #468.

MONAI's `SupervisedTrainer` is an Ignite engine, so this adds a MONAI handler.
Pass `TraceMLHandler()` in `train_handlers` and each training step is timed from
the engine's own events.

```python
from traceml_ai.integrations import monai as traceml_monai

traceml_monai.init()
trainer = SupervisedTrainer(..., train_handlers=[traceml_monai.TraceMLHandler()])
```

## What lands

- `traceml_ai.integrations.monai` with `init()` and `TraceMLHandler`.
- One TraceML step is one optimizer update, as in the Lightning callback, so a
  step spans several iterations under `accumulation_steps`.
- Input Wait comes from the engine's `GET_BATCH_STARTED` to
  `GET_BATCH_COMPLETED` bracket, timed on the training thread. `init()` leaves
  the torch DataLoader fetch patch off, so a loader that fetches on a background
  thread, such as `ThreadDataLoader`, is timed by what the loop waited rather
  than by what its producer did (#463).
- Step time, step memory, and H2D on CUDA.
- Only a `SupervisedTrainer` that runs `SupervisedTrainer._iteration` unchanged
  is traced, including a subclass that inherits it. `GanTrainer`,
  `AdversarialTrainer`, evaluators, `iteration_update=` and an overridden
  `_iteration` each get one warning and nothing attached.
- A `monai` extra, and a `test-monai` CI job that also reruns the file at the
  declared Ignite floor.

Forward, backward and optimizer time are #465. The guide, the support-matrix row
and the example are #466, and the CHANGELOG entry says so.

## Evidence

GATE: ran=true. `pytest tests/integrations/test_monai.py` is 43 passed on MONAI
1.6.0 with Ignite 0.5.5. The `test-monai` job on this head builds Python 3.10
with torch 2.14.0 and runs the file twice, once at the pinned versions and once
at the declared floor Ignite 0.4.11: 43 passed both times. `pytest
tests/integrations` is 264 passed, 8 skipped. The full suite is 1954 passed, 8
skipped, against 1911 passed, 8 skipped on `version_0.4.1` at 2dc86368. The 43
tests this branch adds are the whole difference.

TRANSITIONS: tested as A to B, not as single states. Not tracing to tracing to
cleanup, once per run. Group open to group published, and group open to group
dropped at a crash or an unfinished epoch. Iteration with a step to iteration
without one, which is a `prepare_batch` replaced mid-run and now warns.
Recording to draining when the step budget is spent. Kill switch off to on
during a run. First attach to second attach refused, per handler and per
trainer. Fetch start to fetch recorded, and fetch start to fetch dropped, which
is the unmatched start Ignite leaves at the end of a length-less loader's first
epoch. Capture owned to capture detached by another producer.

RANKS: not exercised and not claimed. The handler has no per-rank branch. Step
numbers come from the existing process-local session counter, so each rank
publishes its own exactly as the other integrations do. `multi_process` and
`multi_node` stay unclaimed until the support-matrix row lands in #466.

TRIPWIRE: every guard is fired on purpose by a test. The unsupported-engine
warning across 8 engine types, the refusal when a DataLoader fetch patch is
already installed, the H2D-not-armed warning on a CUDA engine, the mid-run
`prepare_batch` replacement, the second-handler refusal, a group restarted by
another producer, and an internal error forced at run start, at step open and
at run end. A mutation pass removed each promised behaviour in turn: 25 of 25
mutations were caught by at least one test, with the unmutated control green.

TWINS: searched the concept, not a symbol. The concept is a timing region opened
in one callback and closed in another, which is where an orphaned region comes
from. Other sites are `lightning.py:239`, `:300` and `huggingface.py:180`,
`:192`, and both are bounded by hooks their framework always pairs. Input Wait
is produced only by `dataloader_patch.py:131` and `wrappers.py:95`, each inside
a single `with`. No other integration reads a framework fetch bracket, so the
unmatched-start case has no twin.

End to end under `traceml run`, 32 iterations with `accumulation_steps=2` on a
`ThreadDataLoader` that cannot keep up: the summary reports 16 steps for the
trainer's 16 optimizer updates, and Input Wait of 120.53 ms per step, 1.9285 s
over the run, against the 1.9291 s the loop measured for itself.

## Review

An external model and five internal review passes produced 32 findings. Fixed
here: a test-isolation defect where a DataLoader patch left installed by an
earlier integration test made this file fail in a full-directory run; a guard
that was per handler instead of per trainer, so two handler instances doubled
every number silently; tests that counted events without ever reading a clock;
a fetch clock that kept taking CUDA events after the step budget was spent; a
`prepare_batch` replaced mid-run that stopped tracing in silence; a fail-open
read of `accumulation_steps`; a module-level `import torch`; an unpinned CI gate
whose declared Ignite floor was never run; and a group that could lose
measurements when another producer detached the shared capture. The rest were
refuted against sibling behaviour in `lightning.py` and `huggingface.py` or
filed as follow-ups.

## Summary by CodeRabbit

* **New Features**
  * Added MONAI integration for `SupervisedTrainer` workflows.
  * Captures optimizer-update steps, step duration, input wait time, memory usage, and CUDA host-to-device transfer timing.
  * Supports accumulation-step grouping and safely handles interrupted or incomplete captures.
  * Added optional MONAI dependencies for installation.

* **Documentation**
  * Added an unreleased changelog entry describing MONAI support and captured metrics.

* **Tests**
  * Added comprehensive integration coverage for MONAI training scenarios and supported loader configurations.

### CODE REVIEW

**Comment by abhinavsriva** [comment]
@Pendu: The overall design looks appropriate for Part 1. Ignite’s fetch events measure the training thread’s actual wait, and no shared PyTorch DataLoader patch change is needed. The dedicated CPU integration job and Ignite-floor check also make sense.

Two small changes:
1. Fix unfinished accumulation crossing epochs. With a length-less loader, four batches per epoch, accumulation_steps=3, and two epochs, TraceML reports groups of [3, 4, 1] batches instead of [3, 3, 1]. MONAI clears the unfinished first-epoch gradients, but TraceML carries their measurements into the next update. Please discard that unfinished capture at the epoch boundary and add this regression test. This concerns measurement attribution; training itself is unaffected.

2. Correct the CI comment. It says test-latest-deps provides MONAI compatibility coverage, but that job does not install or test MONAI. Correcting the comment is enough here; expanded scheduled coverage can follow separately.

The planned phase timings, documentation, support matrix, and examples can remain in their follow-up PRs. I do not see a need for a broader redesign before accepting Part 1 once these points are addressed.

**Review by abhinavsriva — CHANGES_REQUESTED** [review]
Minor change is requested

**Comment by Pendu** [comment]
Both points addressed.

1. Fixed. The rule you traced is right.
MONAI zeroes an unstepped window's gradients (`should_zero_grad`) once it learns the epoch length.
The old code never read that signal, so it kept merging the leftover iteration into whatever group stepped next.
`_zeroes_gradients` now mirrors `should_zero_grad`, the same way `_ends_update_group` mirrors `should_step`.
It runs on `GET_BATCH_STARTED`, the earliest hook for a new iteration, so the dropped window's own fetch measurement isn't lost along with it.
Groups are `[3, 3, 1]` now, with a warning fired once per run.
Added `test_an_epoch_boundary_mid_accumulation_drops_the_unstepped_group`, plus a mutation proving the guard binds.

2. Fixed. The comment now says `test-latest-deps` doesn't install MONAI, rather than implying it does.

A second review pass, with real repo access this time, found two more gaps in that fix and one in the existing handler.
The discard ran outside the file's own best-effort boundary, so an exception in it (an empty loader with `epoch_length == 0`, for one) would have reached training.
Reproduced the crash on the unwrapped code and confirmed the fix survives it, both under a real `traceml run`.
The accumulation tests asserted group counts and sizes but never which iterations MONAI actually stepped on, so a handler whose rule quietly drifted from MONAI's could still pass.
Every row now asserts the exact iteration list.
A `zero_grad` spy proves MONAI really does wipe the window at iteration 5 without stepping on it, which is the one premise the whole fix rests on.
Also added the missing negative control: a loader whose windows always step cleanly, proving the discard never fires when there is nothing stale.

While tracing the fix I found the mutation runner had two anchors that collided with new code.
Two module-level lines now appear in two functions instead of one.
Fixed those too so the mutation pass still discriminates correctly.

Also picked up CodeRabbit's import-guard suggestion: `except Exception` instead of `except ImportError`, so a broken MONAI install still sets `IS_MONAI_AVAILABLE = False` instead of crashing the import.

Re-verified: 47 passed in `test_monai.py`, 268 passed / 8 skipped in `tests/integrations`, 1958 passed / 8 skipped full suite.
27 of 27 mutations killed, control green.

**Review by abhinavsriva — APPROVED** [review]
Re-reviewed Part 1. Both requested changes are addressed, including regression coverage for unfinished accumulation. I found no further blocking issue within this PR’s scope. The changelog conflict is resolved locally, and 179 targeted tests pass. Once the resolution is pushed and CI is green, this looks ready to merge.
